File: dsa/12-graph/dfsGraph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

void dfs(int node, vector<int> adj[], unordered_map<int, bool> &visited, vector<int> &ls)
{
    visited[node] = 1;
    cout << node << " ";
    for (auto neighbor : adj[node])
    {
        if (!visited[neighbor])
        {
            dfs(neighbor, adj, visited, ls);
        }
    }
}

void addEdge(vector<int> adj[], int s, int d)
{
    adj[s].push_back(d);
    adj[d].push_back(s);
}

int main()
{
    int V = 5;
    vector<int> adj[V];
    unordered_map<int, bool> visited;
    vector<int> ls;

    addEdge(adj, 0, 1);
    addEdge(adj, 0, 2);
    addEdge(adj, 1, 2);
    addEdge(adj, 2, 0);
    addEdge(adj, 2, 3);
    addEdge(adj, 3, 3);

    dfs(2, adj, visited, ls);

    return 0;
}

File: dsa/12-graph/dfsGraph.py
class Graph:
    def __init__(self, size) -> None:
        self.size = size
        self.adjList = [[] for _ in range(size)]
        self.visit = set()

    def addEdge(self, u, v):
        self.adjList[u].append(v)
        self.adjList[v].append(u)

    def dfs(self, node):
        self.visit.add(node)
        for i in self.adjList[node]:
            if i not in self.visit:
                self.dfs(i)

    def traverseGraph(self):
        for d in self.visit:
            print(f"{d}", end=" ")
        print()


if __name__ == "__main__":
    g = Graph(4)
    g.addEdge(0, 1)
    g.addEdge(0, 2)
    g.addEdge(1, 2)
    g.addEdge(2, 0)
    g.addEdge(2, 3)
    g.addEdge(3, 3)
    g.dfs(2)
    g.traverseGraph()

'''
TC: O(E+V) V is the number of vertices and E is the number of edges in the graph
SC: O(V + E), visited array size V is required, And stack size=E
'''